# Objective
* The goal of this project is to predict customer churn using various machine learning models and choosing the best model that gives the most accurate predictions.
* Customer attrition, also known as customer churn, customer turnover, or customer defection, is the loss of clients or customers.
* Each row represents a customer, each column contains a customer’s attribute.

# Data collected
* Download the data files.
  * The data is available here: https://www.kaggle.com/blastchar/telco-customer-churn/download

In [3]:
%sh
wget -O customer_churn https://www.dropbox.com/s/xct68iza4c8v7z2/WA_Fn-UseC_-Telco-Customer-Churn1.csv?dl=0

--2019-12-13 23:31:16-- https://www.dropbox.com/s/xct68iza4c8v7z2/WA_Fn-UseC_-Telco-Customer-Churn1.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.1.1, 2620:100:6016:1::a27d:101
Connecting to www.dropbox.com (www.dropbox.com)|162.125.1.1|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/xct68iza4c8v7z2/WA_Fn-UseC_-Telco-Customer-Churn1.csv [following]
--2019-12-13 23:31:16-- https://www.dropbox.com/s/raw/xct68iza4c8v7z2/WA_Fn-UseC_-Telco-Customer-Churn1.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://ucaaa9470e686e63347ed83701d7.dl.dropboxusercontent.com/cd/0/inline/AuP2ngZlP4RmXkGSX7aXQVy3EWUFgiED-TqwgiiD_cbPOg2dz-Vk4fbRbUOYBpI0LMDdkdYu8BqA2GpUzgZhJMtpWRc6siCPSzslWkuHcUIWoV-f4KjOxqB8BIHRwPi1rL0/file# [following]
--2019-12-13 23:31:16-- https://ucaaa9470e686e63347ed83701d7.dl.dropboxusercontent.com/cd/0/inline/AuP2ngZlP4RmXkGSX7aXQVy3EWUFgiED-TqwgiiD_cbPOg2dz-Vk4fbRbUOYBpI0LMDdkdYu8BqA2GpUzgZhJMtpWRc6siCPSzslWkuHcUIWoV-f4KjOxqB8BIHRwPi1rL0/file
Resolving ucaaa9470e686e63347ed83701d7.dl.dropboxusercontent.com (ucaaa9470e686e63347ed83701d7.dl.dropboxusercontent.com)... 162.125.1.6, 2620:100:6016:6::a27d:106
Connecting to ucaaa9470e686e63347ed83701d7.dl.dropboxusercontent.com (ucaaa9470e686e63347ed83701d7.dl.dropboxusercontent.com)|162.125.1.6|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 977523 (955K) [text/plain]
Saving to: ‘customer_churn’

 0K .......... .......... .......... .......... .......... 5% 2.97M 0s
 50K .......... .......... .......... .......... .......... 10% 11.7M 0s
 100K .......... .......... .......... .......... .......... 15% 11.5M 0s
 150K .......... .......... .......... .......... .......... 20% 13.0M 0s
 200K .......... .......... .......... .......... .......... 26% 32.3M 0s
 250K .......... .......... .......... .......... .......... 31% 17.5M 0s
 300K .......... .......... .......... .......... .......... 36% 59.2M 0s
 350K .......... .......... .......... .......... .......... 41% 17.7M 0s
 400K .......... .......... .......... .......... .......... 47% 79.1M 0s
 450K .......... .......... .......... .......... .......... 52% 28.0M 0s
 500K .......... .......... .......... .......... .......... 57% 46.6M 0s
 550K .......... .......... .......... .......... .......... 62% 73.7M 0s
 600K .......... .......... .......... .......... .......... 68% 44.3M 0s
 650K .......... .......... .......... .......... .......... 73% 61.0M 0s
 700K .......... .......... .......... .......... .......... 78% 106M 0s
 750K .......... .......... .......... .......... .......... 83% 57.6M 0s
 800K .......... .......... .......... .......... .......... 89% 53.2M 0s
 850K .......... .......... .......... .......... .......... 94% 82.5M 0s
 900K .......... .......... .......... .......... .......... 99% 82.0M 0s
 950K .... 100% 74.2M=0.05s

2019-12-13 23:31:17 (20.3 MB/s) - ‘customer_churn’ saved [977523/977523]

In [4]:
customer_churn = spark.read.csv('file:/databricks/driver/customer_churn', inferSchema=True, header=True, mode='DROPMALFORMED')

In [5]:
#displaying fisrt five rows of the dataframe
customer_churn.show(5)

+----------+------+-------------+-------+----------+------+------------+----------------+---------------+--------------+------------+----------------+-----------+-----------+---------------+--------------+----------------+--------------------+--------------+------------+-----+
customerID|gender|SeniorCitizen|Partner|Dependents|tenure|PhoneService| MultipleLines|InternetService|OnlineSecurity|OnlineBackup|DeviceProtection|TechSupport|StreamingTV|StreamingMovies| Contract|PaperlessBilling| PaymentMethod|MonthlyCharges|TotalCharges|Churn|
+----------+------+-------------+-------+----------+------+------------+----------------+---------------+--------------+------------+----------------+-----------+-----------+---------------+--------------+----------------+--------------------+--------------+------------+-----+
7590-VHVEG|Female| 0| Yes| No| 1| No|No phone service| DSL| No| Yes| No| No| No| No|Month-to-month| Yes| Electronic check| 29.85| 29.85| No|
5575-GNVDE| Male| 0| No| No| 34| Yes| No| DSL| Yes| No| Yes| No| No| No| One year| No| Mailed check| 56.95| 1889.5| No|
3668-QPYBK| Male| 0| No| No| 2| Yes| No| DSL| Yes| Yes| No| No| No| No|Month-to-month| Yes| Mailed check| 53.85| 108.15| Yes|
7795-CFOCW| Male| 0| No| No| 45| No|No phone service| DSL| Yes| No| Yes| Yes| No| No| One year| No|Bank transfer (au...| 42.3| 1840.75| No|
9237-HQITU|Female| 0| No| No| 2| Yes| No| Fiber optic| No| No| No| No| No| No|Month-to-month| Yes| Electronic check| 70.7| 151.65| Yes|
+----------+------+-------------+-------+----------+------+------------+----------------+---------------+--------------+------------+----------------+-----------+-----------+---------------+--------------+----------------+--------------------+--------------+------------+-----+
only showing top 5 rows

In [6]:
customer_churn.printSchema()

root
-- customerID: string (nullable = true)
-- gender: string (nullable = true)
-- SeniorCitizen: integer (nullable = true)
-- Partner: string (nullable = true)
-- Dependents: string (nullable = true)
-- tenure: integer (nullable = true)
-- PhoneService: string (nullable = true)
-- MultipleLines: string (nullable = true)
-- InternetService: string (nullable = true)
-- OnlineSecurity: string (nullable = true)
-- OnlineBackup: string (nullable = true)
-- DeviceProtection: string (nullable = true)
-- TechSupport: string (nullable = true)
-- StreamingTV: string (nullable = true)
-- StreamingMovies: string (nullable = true)
-- Contract: string (nullable = true)
-- PaperlessBilling: string (nullable = true)
-- PaymentMethod: string (nullable = true)
-- MonthlyCharges: double (nullable = true)
-- TotalCharges: double (nullable = true)
-- Churn: string (nullable = true)

# Data Explained
* KEY FIELDS:
  * customerID: Customer ID
  * gender: Whether the customer is a male or a female
  * SeniorCitizen: Whether the customer is a senior citizen or not (1, 0)
  * Partner: Whether the customer has a partner or not (Yes, No)
  * Dependents: Whether the customer has dependents or not (Yes, No)
  * tenure: Number of months the customer has stayed with the company
  * PhoneService: Whether the customer has a phone service or not (Yes, No)
  * MultipleLines: Whether the customer has multiple lines or not (Yes, No, No phone service)
  * InternetService: Customer’s internet service provider (DSL, Fiber optic, No)
  * OnlineSecurity: Whether the customer has online security or not (Yes, No, No internet service)
  * OnlineBackup: Whether the customer has online backup or not (Yes, No, No internet service)
  * DeviceProtection: Whether the customer has device protection or not (Yes, No, No internet service)
  * TechSupport: Whether the customer has tech support or not (Yes, No, No internet service)
  * StreamingTV: Whether the customer has streaming TV or not (Yes, No, No internet service)
  * StreamingMovies: Whether the customer has streaming movies or not (Yes, No, No internet service)
  * Contract: The contract term of the customer (Month-to-month, One year, Two year)
  * PaperlessBilling: Whether the customer has paperless billing or not (Yes, No)
  * PaymentMethod: The customer’s payment method (Electronic check, Mailed check, Bank transfer (automatic), Credit card (automatic))
  * MonthlyCharges: The amount charged to the customer monthly
  * TotalCharges: The total amount charged to the customer
* PREDICTOR FIELD:
  * Churn: Whether the customer churned or not (Yes or No)

In [8]:
# Displaying the stats 
display(customer_churn.summary())

summary customerID gender SeniorCitizen Partner Dependents tenure PhoneService MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies Contract PaperlessBilling PaymentMethod MonthlyCharges TotalCharges Churn count 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 7043 mean null null 0.1621468124378816 null null 32.37114865824223 null null null null null null null null null null null null 64.76169246059922 NaN null stddev null null 0.3686116056100135 null null 24.559481023094442 null null null null null null null null null null null null 30.090047097678482 NaN null min 0002-ORFBO Female 0 No No 0 No No DSL No No No No No No Month-to-month No Bank transfer (automatic) 18.25 18.8 No 25% null null 0 null null 9 null null null null null null null null null null null null 35.5 401.95 null 50% null null 0 null null 29 null null null null null null null null null null null null 70.35 1400.55 null 75% null null 1 null null 72 null null null null null null null null null null null null 118.75 NaN null max 9995-HOTOH Male 1 Yes Yes 72 Yes Yes No Yes Yes Yes Yes Yes Yes Two year Yes Mailed check 118.75 NaN Yes

# Data Cleaning
* Rows containing null values have been dropped to make our data suitable for applying various models.

In [10]:
#Dropping rows with NaN values
print("rows: {}".format(customer_churn.count()))
customer_churn = customer_churn.dropna()
print("rows after dropna",format(customer_churn.count()))

rows: 7043
rows after dropna 7032

# Data Visualization

In [12]:
# Plotting the distribution of customer Churn
import seaborn as sns
import matplotlib.pyplot as plt

from matplotlib.pyplot import figure


p_df = customer_churn.toPandas()

sns.countplot(x='Churn',data=p_df).set_title('Distribution of Churn')

fig = plt.gcf()
display(fig)

In [13]:
# The plot shows that variation of Churn based on MonthlyCharges and TotalCharges
sns.scatterplot(x='MonthlyCharges',y='TotalCharges',hue='Churn',data=p_df).set_title('MonthlyCharges vs TotalCharges with respect to Churn')
fig = plt.gcf()
display(fig)

In [14]:
#Plot shows that if the contract is for a longer period of time, the customer churn is less.
sns.scatterplot(x='Contract',y='TotalCharges',hue='Churn',data=p_df).set_title('Variation of Churn and Contract')
fig = plt.gcf()
display(fig)

In [15]:
#Plot shows the variation of Churn with respect to PaymentMethod
sns.swarmplot(x='PaymentMethod',y='MonthlyCharges',hue='Churn',data=p_df).set_title('variation of Churn with respect to PaymentMethod')
fig = plt.gcf()
fig.set_size_inches(9, 7)
display(fig)

# Data Modeling

In [17]:
from pyspark.ml.feature import StringIndexer,VectorAssembler
from pyspark.ml.classification import LogisticRegression, RandomForestClassifier, GBTClassifier, NaiveBayes
from pyspark.ml import Pipeline,Model
from pyspark.ml.tuning import ParamGridBuilder

stringIndexer_1= StringIndexer(inputCol="gender", outputCol="gender_IX")
stringIndexer_2 = StringIndexer(inputCol="Partner", outputCol="Partner_IX")
stringIndexer_3 = StringIndexer(inputCol="Dependents", outputCol="Dependents_IX")
stringIndexer_4 = StringIndexer(inputCol="PhoneService", outputCol="PhoneService_IX")
stringIndexer_5 = StringIndexer(inputCol="MultipleLines", outputCol="MultipleLines_IX")
stringIndexer_6 = StringIndexer(inputCol="InternetService", outputCol="InternetService_IX")
stringIndexer_7 = StringIndexer(inputCol="OnlineSecurity", outputCol="OnlineSecurity_IX")
stringIndexer_8 = StringIndexer(inputCol="OnlineBackup", outputCol="OnlineBackup_IX")
stringIndexer_9 = StringIndexer(inputCol="DeviceProtection", outputCol="DeviceProtection_IX")
stringIndexer_10 = StringIndexer(inputCol="TechSupport", outputCol="TechSupport_IX")
stringIndexer_11= StringIndexer(inputCol="StreamingTV", outputCol="StreamingTV_IX")
stringIndexer_12= StringIndexer(inputCol="StreamingMovies", outputCol="StreamingMovies_IX")
stringIndexer_13= StringIndexer(inputCol="Contract", outputCol="Contract_IX")
stringIndexer_14= StringIndexer(inputCol="PaperlessBilling", outputCol="PaperlessBilling_IX")
stringIndexer_15= StringIndexer(inputCol="PaymentMethod", outputCol="PaymentMethod_IX")

vectorAssembler_features = VectorAssembler(inputCols=["gender_IX", "Partner_IX","Dependents_IX","PhoneService_IX","MultipleLines_IX","InternetService_IX","OnlineSecurity_IX","OnlineBackup_IX","DeviceProtection_IX","TechSupport_IX","StreamingTV_IX","StreamingMovies_IX","Contract_IX","PaperlessBilling_IX","PaymentMethod_IX","MonthlyCharges","TotalCharges","tenure"],outputCol="features")
stringIndexer_label = StringIndexer(inputCol="Churn", outputCol="label").fit(customer_churn)

# bestmodel_features = VectorAssembler(inputCols=[ "OnlineSecurity_IX","TechSupport_IX","Contract_IX","PaperlessBilling_IX","tenure"], outputCol="features")

# lr = LogisticRegression(labelCol= "label", featuresCol = "features", maxIter = 10, regParam = 0.0001, elasticNetParam = 1.0)
# rf = RandomForestClassifier(numTrees=50)

pipeline = Pipeline(stages=[])

basePipeline = [stringIndexer_1, stringIndexer_2,stringIndexer_3, stringIndexer_4,stringIndexer_5, stringIndexer_6,stringIndexer_7, stringIndexer_8,stringIndexer_9, stringIndexer_10,stringIndexer_11, stringIndexer_12,stringIndexer_13, stringIndexer_14,stringIndexer_15,stringIndexer_label, vectorAssembler_features]

# basePipeline_bestmodel = [stringIndexer_1, stringIndexer_2,stringIndexer_3, stringIndexer_4,stringIndexer_5, stringIndexer_6,stringIndexer_7, stringIndexer_8,stringIndexer_9, stringIndexer_10,stringIndexer_11, stringIndexer_12,stringIndexer_13, stringIndexer_14,stringIndexer_15,stringIndexer_label, bestmodel_features]

lr = LogisticRegression(maxIter=10)
pl_lr = basePipeline + [lr]
pg_lr = ParamGridBuilder()\
          .baseOn({pipeline.stages: pl_lr})\
          .addGrid(lr.regParam,[0.01, .04])\
          .addGrid(lr.elasticNetParam,[0.1, 0.4])\
          .build()

rf = RandomForestClassifier(numTrees=50)
pl_rf = basePipeline + [rf]
pg_rf = ParamGridBuilder()\
      .baseOn({pipeline.stages: pl_rf})\
      .build()

gb = GBTClassifier(labelCol="label", featuresCol="features",  maxBins=32,  maxDepth=10, maxIter=15)
pl_gb = basePipeline + [gb]
pg_gb = ParamGridBuilder()\
      .baseOn({pipeline.stages: pl_gb})\
      .build()

nb = NaiveBayes()
pl_nb = basePipeline + [nb]
pg_nb = ParamGridBuilder()\
      .baseOn({pipeline.stages: pl_nb})\
      .addGrid(nb.smoothing,[0.4,1.0])\
      .build()


paramGrid = pg_lr + pg_rf + pg_gb + pg_nb


In [18]:
splitted_data = customer_churn.randomSplit([0.6, 0.4], 24)   
train_data = splitted_data[0]
test_data = splitted_data[1]

print("Number of training records: " + str(train_data.count()))
print("Number of testing records : " + str(test_data.count()))

Number of training records: 4203
Number of testing records : 2829

In [19]:
%sh
pip install mlflow

Requirement already satisfied: mlflow in /databricks/python3/lib/python3.7/site-packages (1.4.0)
Requirement already satisfied: pandas in /databricks/python3/lib/python3.7/site-packages (from mlflow) (0.24.2)
Requirement already satisfied: querystring-parser in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.2.4)
Requirement already satisfied: protobuf>=3.6.0 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (3.11.1)
Requirement already satisfied: simplejson in /databricks/python3/lib/python3.7/site-packages (from mlflow) (3.17.0)
Requirement already satisfied: sqlalchemy in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.3.11)
Requirement already satisfied: python-dateutil in /databricks/python3/lib/python3.7/site-packages (from mlflow) (2.8.0)
Requirement already satisfied: gitpython>=2.1.0 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (3.0.5)
Requirement already satisfied: Flask in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.1.1)
Requirement already satisfied: docker>=4.0.0 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (4.1.0)
Requirement already satisfied: gorilla in /databricks/python3/lib/python3.7/site-packages (from mlflow) (0.3.0)
Requirement already satisfied: six>=1.10.0 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.12.0)
Requirement already satisfied: pyyaml in /databricks/python3/lib/python3.7/site-packages (from mlflow) (5.2)
Requirement already satisfied: click>=7.0 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (7.0)
Requirement already satisfied: databricks-cli>=0.8.7 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (0.9.1)
Requirement already satisfied: sqlparse in /databricks/python3/lib/python3.7/site-packages (from mlflow) (0.3.0)
Requirement already satisfied: gunicorn; platform_system != "Windows" in /databricks/python3/lib/python3.7/site-packages (from mlflow) (20.0.4)
Requirement already satisfied: numpy in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.16.2)
Requirement already satisfied: alembic in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.3.1)
Requirement already satisfied: entrypoints in /databricks/python3/lib/python3.7/site-packages (from mlflow) (0.3)
Requirement already satisfied: cloudpickle in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.2.2)
Requirement already satisfied: requests>=2.17.3 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (2.21.0)
Requirement already satisfied: pytz>=2011k in /databricks/python3/lib/python3.7/site-packages (from pandas->mlflow) (2018.9)
Requirement already satisfied: setuptools in /usr/lib/python3.7/site-packages (from protobuf>=3.6.0->mlflow) (40.8.0)
Requirement already satisfied: gitdb2>=2.0.0 in /databricks/python3/lib/python3.7/site-packages (from gitpython>=2.1.0->mlflow) (2.0.6)
Requirement already satisfied: Jinja2>=2.10.1 in /databricks/python3/lib/python3.7/site-packages (from Flask->mlflow) (2.10.3)
Requirement already satisfied: Werkzeug>=0.15 in /databricks/python3/lib/python3.7/site-packages (from Flask->mlflow) (0.16.0)
Requirement already satisfied: itsdangerous>=0.24 in /databricks/python3/lib/python3.7/site-packages (from Flask->mlflow) (1.1.0)
Requirement already satisfied: websocket-client>=0.32.0 in /databricks/python3/lib/python3.7/site-packages (from docker>=4.0.0->mlflow) (0.56.0)
Requirement already satisfied: configparser>=0.3.5 in /databricks/python3/lib/python3.7/site-packages (from databricks-cli>=0.8.7->mlflow) (4.0.2)
Requirement already satisfied: tabulate>=0.7.7 in /databricks/python3/lib/python3.7/site-packages (from databricks-cli>=0.8.7->mlflow) (0.8.6)
Requirement already satisfied: Mako in /databricks/python3/lib/python3.7/site-packages (from alembic->mlflow) (1.1.0)
Requirement already satisfied: python-editor>=0.3 in /databricks/python3/lib/python3.7/site-packages (from alembic->mlflo

In [20]:
from pyspark.ml.evaluation import BinaryClassificationEvaluator
from pyspark.ml.tuning import CrossValidator
cv = CrossValidator()\
      .setEstimator(pipeline)\
      .setEvaluator(BinaryClassificationEvaluator())\
      .setEstimatorParamMaps(paramGrid)\
      .setNumFolds(2)

cvmodel = cv.fit(train_data)

/databricks/spark/python/pyspark/ml/util.py:791: UserWarning: Can not find mlflow. To enable mlflow logging, install MLflow library from PyPi.
 warnings.warn(_MLflowInstrumentation._NO_MLFLOW_WARNING)

In [21]:
cvmodel.bestModel.stages

Out[250]: [StringIndexer_2e1f2903b61f,
 StringIndexer_b6d47e99904a,
 StringIndexer_547eb7eae248,
 StringIndexer_3ed4b21feed1,
 StringIndexer_76ce0e9c176f,
 StringIndexer_52893edc800c,
 StringIndexer_a8ff2249891b,
 StringIndexer_fd6a04867208,
 StringIndexer_e8d42f22f2bd,
 StringIndexer_2f9847285247,
 StringIndexer_0b3e55f86220,
 StringIndexer_31f281176301,
 StringIndexer_25fddd2b0c9a,
 StringIndexer_a79da00412a1,
 StringIndexer_beffcd678f1e,
 StringIndexer_130dc57ca15d,
 VectorAssembler_bbfc72489d97,
 RandomForestClassificationModel (uid=RandomForestClassifier_07ca609077e5) with 50 trees]

In [22]:
import pandas as pd
model = pd.DataFrame(cvmodel.bestModel.stages[-1].featureImportances.toArray(), columns=["values"])
features = ["gender_IX", "Partner_IX","Dependents_IX","PhoneService_IX","MultipleLines_IX","InternetService_IX","OnlineSecurity_IX","OnlineBackup_IX","DeviceProtection_IX","TechSupport_IX","StreamingTV_IX","StreamingMovies_IX","Contract_IX","PaperlessBilling_IX","PaymentMethod_IX","MonthlyCharges","TotalCharges","tenure"]
features_col = pd.Series(features)
model["features"] = features_col
model

values 
 features 
 
 
 
 
 0 
 0.000752 
 gender_IX 
 
 
 1 
 0.002164 
 Partner_IX 
 
 
 2 
 0.002537 
 Dependents_IX 
 
 
 3 
 0.001144 
 PhoneService_IX 
 
 
 4 
 0.003478 
 MultipleLines_IX 
 
 
 5 
 0.072352 
 InternetService_IX 
 
 
 6 
 0.096543 
 OnlineSecurity_IX 
 
 
 7 
 0.020776 
 OnlineBackup_IX 
 
 
 8 
 0.019097 
 DeviceProtection_IX 
 
 
 9 
 0.144378 
 TechSupport_IX 
 
 
 10 
 0.011941 
 StreamingTV_IX 
 
 
 11 
 0.016971 
 StreamingMovies_IX 
 
 
 12 
 0.281917 
 Contract_IX 
 
 
 13 
 0.008951 
 PaperlessBilling_IX 
 
 
 14 
 0.051006 
 PaymentMethod_IX 
 
 
 15 
 0.052215 
 MonthlyCharges 
 
 
 16 
 0.088612 
 TotalCharges 
 
 
 17 
 0.125168 
 tenure

In [23]:
import numpy as np
cvmodel.getEstimatorParamMaps()[ np.argmax(cvmodel.avgMetrics) ]

Out[252]: {Param(parent='Pipeline_2ca0c7706100', name='stages', doc='a list of pipeline stages'): [StringIndexer_2e1f2903b61f,
 StringIndexer_b6d47e99904a,
 StringIndexer_547eb7eae248,
 StringIndexer_3ed4b21feed1,
 StringIndexer_76ce0e9c176f,
 StringIndexer_52893edc800c,
 StringIndexer_a8ff2249891b,
 StringIndexer_fd6a04867208,
 StringIndexer_e8d42f22f2bd,
 StringIndexer_2f9847285247,
 StringIndexer_0b3e55f86220,
 StringIndexer_31f281176301,
 StringIndexer_25fddd2b0c9a,
 StringIndexer_a79da00412a1,
 StringIndexer_beffcd678f1e,
 StringIndexer_130dc57ca15d,
 VectorAssembler_bbfc72489d97,
 RandomForestClassifier_07ca609077e5]}

In [24]:
import re
def paramGrid_model_name(model):
  params = [v for v in model.values() if type(v) is not list]
  name = [v[-1] for v in model.values() if type(v) is list][0]
  name = re.match(r'([a-zA-Z]*)', str(name)).groups()[0]
  return "{}{}".format(name,params)

# Resulting metric and model description
# get the measure from the CrossValidator, cvModel.avgMetrics
# get the model name & params from the paramGrid
# put them together here:
kmeans_measures = zip(cvmodel.avgMetrics, [paramGrid_model_name(m) for m in paramGrid])
metrics,model_names = zip(*kmeans_measures)

In [25]:
import seaborn as sns
import matplotlib.pyplot as plt

plt.clf() # clear figure
fig = plt.figure( figsize=(5, 5))
plt.style.use('fivethirtyeight')
axis = fig.add_axes([0.1, 0.3, 0.8, 0.6])
# plot the metrics as Y
#plt.plot(range(len(model_names)),metrics)
plt.bar(range(len(model_names)),metrics)
# plot the model name & param as X labels
plt.xticks(range(len(model_names)), model_names, rotation=70, fontsize=6)
plt.yticks(fontsize=6)
#plt.xlabel('model',fontsize=8)
plt.ylabel('ROC AUC (better is greater)',fontsize=8)
plt.title('Model evaluations')
display(plt.show())

In [26]:
stringIndexer_7= StringIndexer(inputCol="OnlineSecurity", outputCol="OnlineSecurity_IX")
stringIndexer_10 = StringIndexer(inputCol="TechSupport", outputCol="TechSupport_IX")
stringIndexer_13 = StringIndexer(inputCol="Contract", outputCol="Contract_IX")
stringIndexer_15 = StringIndexer(inputCol="PaymentMethod", outputCol="PaymentMethod_IX")
stringIndexer_6 = StringIndexer(inputCol="InternetService", outputCol="InternetService_IX")
stringIndexer_8 = StringIndexer(inputCol="OnlineBackup", outputCol="OnlineBackup_IX")
stringIndexer_9 = StringIndexer(inputCol="DeviceProtection", outputCol="DeviceProtection_IX")
stringIndexer_11 = StringIndexer(inputCol="StreamingTV", outputCol="StreamingTV_IX")
stringIndexer_12 = StringIndexer(inputCol="StreamingMovies", outputCol="StreamingMovies_IX")
stringIndexer_label = StringIndexer(inputCol="Churn", outputCol="label").fit(customer_churn)
# vectorAssembler_features = VectorAssembler(inputCols=["OnlineSecurity_IX","TechSupport_IX","Contract_IX","PaperlessBilling_IX","tenure"],outputCol="features")
bestmodel_features = VectorAssembler(inputCols=[ "InternetService_IX","OnlineBackup_IX","DeviceProtection_IX","StreamingTV_IX","StreamingMovies_IX","PaymentMethod_IX","MonthlyCharges","TotalCharges","OnlineSecurity_IX","TechSupport_IX","Contract_IX","tenure"], outputCol="features")
basePipeline_bm = [stringIndexer_7, stringIndexer_10,stringIndexer_13, stringIndexer_15, stringIndexer_label,stringIndexer_6,stringIndexer_8,stringIndexer_9,stringIndexer_11,stringIndexer_12, bestmodel_features]
rf = RandomForestClassifier(numTrees=50)
# pl_rf = basePipeline_bm + [rf]
basepipe=Pipeline(stages=basePipeline_bm + [rf])
rfModel = basepipe.fit(train_data)
predictions = rfModel.transform(test_data)

# Prediction

###### 
* Predicting churn label based on various features available in the data set using Logistic Regression, RandomForestClassifier, Naive Bayes and gradient Boosted Trees Classifier. The model that gave us the most accurate predictions is the Random Forest Classifier.
* We further improved the Random Forest Classifier by considering feature importance and eliminated few features that did not help in the prediction.
* On the test data the model predicted 2260 labels correctly and 588 incorrectly

In [29]:

# Model quality
predictions = rfModel.transform(test_data)


In [30]:
predictions.select('prediction','label').show()

+----------+-----+
prediction|label|
+----------+-----+
 0.0| 0.0|
 0.0| 0.0|
 1.0| 1.0|
 1.0| 1.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 1.0| 1.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
 0.0| 0.0|
+----------+-----+
only showing top 20 rows

In [31]:
correct = predictions.where("(label = prediction)").count()
incorrect = predictions.where("(label != prediction)").count()

resultDF = sqlContext.createDataFrame([['correct', correct], ['incorrect', incorrect]], ['metric', 'value'])
display(resultDF)

metric value correct 2270 incorrect 559

# Model Evaluation

##### 
* For evaluating the model we calculated the number of predicted values for positive and negative class and compared them with actual labels.

In [34]:
counts = [predictions.where('label=1').count(), predictions.where('prediction=1').count(),
          predictions.where('label=0').count(), predictions.where('prediction=0').count()]
names = ['actual 1', 'predicted 1', 'actual 0', 'predicted 0']
display(sqlContext.createDataFrame(zip(names,counts),['Measure','Value']))

Measure Value actual 1 745 predicted 1 580 actual 0 2084 predicted 0 2249

In [35]:
from pyspark.mllib.evaluation import BinaryClassificationMetrics
# https://spark.apache.org/docs/latest/mllib-evaluation-metrics.html
# Compute raw scores on the test set
predictionAndLabels = predictions.select('prediction','label').rdd

# Instantiate metrics object
metrics = BinaryClassificationMetrics(predictionAndLabels)

# Area under precision-recall curve
print("Area under PR = %s" % metrics.areaUnderPR)

# Area under ROC curve
print("Area under ROC = %s" % metrics.areaUnderROC)

Area under PR = 0.5638922624135633
Area under ROC = 0.7097821046258486

#Result Visualization

In [37]:
correct = predictions.where("(label = prediction)").count()
incorrect = predictions.where("(label != prediction)").count()

resultDF = sqlContext.createDataFrame([['correct', correct], ['incorrect', incorrect]], ['metric', 'value'])
display(resultDF)

metric value correct 2270 incorrect 559

In [38]:
counts = [predictions.where('label=1').count(), predictions.where('prediction=1').count(),
          predictions.where('label=0').count(), predictions.where('prediction=0').count()]
names = ['actual 1', 'predicted 1', 'actual 0', 'predicted 0']
display(sqlContext.createDataFrame(zip(names,counts),['Measure','Value']))

Measure Value actual 1 745 predicted 1 580 actual 0 2084 predicted 0 2249

### Conclusion
* Predicting churn label based on various features available in the data set using Logistic Regression, RandomForestClassifier, Naive Bayes and gradient Boosted Trees Classifier. The model that gave us the most accurate predictions is the Random Forest Classifier.
* We further improved the Random Forest Classifier by considering feature importance and eliminated few features that did not help in the prediction.
* On the test data the model predicted 2270 labels correctly and 559 incorrectly
* For classification of customer churn, we selected the best model(RFC) which had the ROC of 0.71.